### PR TITLE
[#16329] QR code variants

### LINKED ISSUE #16329: Implement all variants for QR code
# Feature

Implement the "Show QR code" option for any given community channel. See [Figma > Sharing for Mobile > Share from Status](https://www.figma.com/file/Nt3KJ8AXm0VPpIRLxaHnw0/Sharing-for-Mobile?type=design&node-id=2345-300688&t=6ebQeYEMqiGEVxn5-0)

## Update:

Implement all variants for the QR code component:
https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=17928-209739&mode=design&t=mOWDt2exmZ2Al72p-4
image

### PR DESCRIPTION
**Pull request by ulisesmac**
fixes #16329

### Summary

This PR adds all the [QR code variants present in figma](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=28187-85679&mode=design&t=uBOjm2jYssWHVSQw-4) and fix the component's API to match figma parameters. Also the Preview screen has been improved to test all the variants.

photo_2023-10-03_16-36-05 photo_2023-10-03_16-36-06 photo_2023-10-03_16-36-10

Additionally,
- Fix a warning in `status-im2.contexts.chat.messages.content.text.view` about `quo.platform`
- The component `wallet-user-avatar` received `:color` to specify the customization color, so changed its API to receive `:customization-color`
- The function `get-account-qr-image-uri` has been deprecated since its result was not as expected in designs, also keep using that implementation would lead to many problems (more about this in the "Review notes" section below.

- We had two implementations to generate a QR: one based in a JS lib, another one from status-go, only the status go  version exists now.
- The node dependency to "qrcode" has been removed.

### Review notes

We are using an implementation based on requesting the QR from status-go and drawing the avatar using our existing avatar components in React Native.

### Testing notes

Since this change affects all the QR codes in the app, they are worth checking out, the locations are:

- Top right QR button:

  image

- Press on username:

  image

- Profile > Syncing > Add button > Reveal sync code
  image

- Less important since the wallet is being rewritten, but:
  wallet tab > Account three dots > Share
  
  image

#### Platforms

- Android
- iOS

### Steps to test
Apart from Review notes already mentioned, the preview screen:
- Quo2 preview > Share > QR code

status: ready

### CODE REVIEW

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (ulisesmac) [inline src/quo2/components/avatars/wallet_user_avatar.cljs:34]: API of `wallet-user-avatar` changed, from `:color` -> `:customization-color`

File: src/status_im2/common/qr_code_viewer/style.cljs
Comment (ulisesmac) [inline src/status_im2/common/qr_code_viewer/style.cljs:1]: Implementation and node dependency removed

File: src/utils/image_server.cljs
Comment (ulisesmac) [inline src/utils/image_server.cljs]: Function now deprecated, also it had a previous use which has been removed so no warnings from this function are thrown

File: src/quo2/components/share/qr_code/view.cljs
Comment (ulisesmac) [inline src/quo2/components/share/qr_code/view.cljs]: Sadly, the `media-server-port` is taken from re-frame :cry:  and we need it to ask QR codes from status-go.

I'd really like to remove this parameter, since it is always required, any suggestion?

AFAIK, we are not allowed to perform re-frame calls inside the component library, but in this case, this component requires status-go to generate the QR, so, is it possible to add an exception and avoid throwing a warning?

Comment (J-Son89) [reply src/quo2/components/share/qr_code/view.cljs]: for re frame connections we can make a connected component in status_im2, imo it can still be nice to have the UI handled in quo2 components but we have this wrapped in somewhere like `src/status_im2/common/qr-codes/view`.

Comment (J-Son89) [reply src/quo2/components/share/qr_code/view.cljs]: could also just move the whole component there too ofc

File: package.json
Comment (J-Son89) [inline package.json:38]: 🙌

File: src/quo2/components/share/qr_code/view.cljs
Comment (ilmotta) [inline src/quo2/components/share/qr_code/view.cljs]: I would prefer if we could keep the QR code component (and the component lib for that matter) less coupled with Status concepts, thus moving the responsibility to generate the image server URL to the caller of `qr-code/view`. We have a few options:

1. The caller passes a fully valid image server URL, ready to be used by `fast-image` (my preferred approach).
2. The caller passes an image component and the `qr-code/view` only worries about styling it inside a container.
3. The caller passes a thunk to `qr-code/view` that's responsible for returning a fully working image server URL.

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (ilmotta) [inline src/quo2/components/avatars/wallet_user_avatar.cljs:31]: Not your PR's fault, but these `f-name` and `l-name` are probably one of the worst abbreviations I've seen for first and last name.

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/share/qr_code.cljs]: FYI: You can pass this style map to `preview-container` in the prop `component-container-style` and then remove the parent `view`.

File: src/utils/image_server.cljs
Comment (ilmotta) [inline src/utils/image_server.cljs]: This function has no usages, why do you want to keep it in the code?

Comment (ulisesmac) [reply src/utils/image_server.cljs]: @ilmotta 

It has an implementation in Status-go, so I'm not sure about just removing this entirely, Do you think it must be removed?

File: src/quo2/components/share/qr_code/view.cljs
Comment (ulisesmac) [reply src/quo2/components/share/qr_code/view.cljs]: @ilmotta 

I really liked @J-Son89 's suggestion, so  this compnent no longer receives `media-server-port` and `url`, instead it receives `qr-image-uri`, which we expect it to be an uri compatible with `fast-image`.

And created  `status-im2.common.qr-codes.view/qr-code` which hgas the `image-server/get-qr-image-uri-for-any-url` call and passes the value to `quo/qr-code`, so in our screens we are just using the common component and we only pass the url string. :+1: 

Thanks for providing alternatives :)

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (ulisesmac) [reply src/quo2/components/avatars/wallet_user_avatar.cljs:31]: I know, and these components need to be refactored to foilow the `:size` value we recently agreed, so there's still room to improve the API

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (ulisesmac) [reply src/status_im2/contexts/quo_preview/share/qr_code.cljs]: Done

**Comment by ulisesmac** [comment]
@J-Son89 @ilmotta 
I solved your comments. Can I please get a second review? 

Thanks!

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (J-Son89) [inline src/quo2/components/avatars/wallet_user_avatar.cljs:45]: nit this doesn't take theme prop, perhaps use
```
(colors/theme-colors 
(colors/custom-color customization-color 50)
(colors/custom-color customization-color 60)
theme)
```
going to properly fix this api soon! :)

File: src/quo2/components/share/qr_code/view.cljs
Comment (J-Son89) [inline src/quo2/components/share/qr_code/view.cljs:23]: are these props defaulted to `true`? weird if so 🫥

File: src/status_im2/contexts/chat/messages/content/text/view.cljs
Comment (J-Son89) [inline src/status_im2/contexts/chat/messages/content/text/view.cljs]: we shouldn't be using `quo.` here 👁️

File: src/status_im2/contexts/quo_preview/avatars/wallet_user_avatar.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/avatars/wallet_user_avatar.cljs:28]: 🌮

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/share/qr_code.cljs]: please update components to use `quo.core` here if you're making adjustments 👍

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/share/qr_code.cljs:98]: this comment is a bit unnecessary imo and suggests something may be wrong here 🤔 
The custom-color function should take care of this so the component in question takes in either a keyword or a hex string and that custom-colors method should handle the resolving 👍

Comment (J-Son89) [reply src/status_im2/contexts/quo_preview/share/qr_code.cljs:98]: same for all of the components using custom colors really, it should be abstracted away by the colors mechanism

**Review by J-Son89 — APPROVED** [review]
left some comments but looks good 👍

**Review by siddarthkay — CHANGES_REQUESTED** [review]
Hi @ulisesmac : Thanks for your hard work on this PR and I would like to block this PR.

I know the existing designs in develop do not match the designs but I have a PR open which addresses those issues : https://github.com/status-im/status-mobile/pull/16257
I know its a bit old ...

We should not move this processing over to client side simply because its much more efficient to have this on status-go side.

By deprecating the method which you intend to do in this PR we are essentially throwing away all of the work I have done into making QR codes via status-go.

The status-go implementation is customisable and we should attempt to move the initials generation and adding emojis on top of QR code from status-go side.

**Comment by ulisesmac** [comment]
> Hi @ulisesmac : Thanks for your hard work on this PR and I would like to block this PR.
> 
> I know the existing designs in develop do not match the designs but I have a PR open which addresses those issues : #16257 I know its a bit old ...

Oh, that PR seems stuck :thinking:

> We should not move this processing over to client side simply because its much more efficient to have this on status-go side.

:thinking:  IMO, unless the rendering takes a significant amount of time, it's ok to have it in CLJS side. the reasons why I think it's ok are:

- All the app components (even the more complex ones) are being rendered using RN
- The part I'm rendering in RN is pretty simple, just a view wrapper and an overlay with an avatar
- Unless we need that performance improvement you mention, I think we shouldn't try to optimize it. BTW, is this performnace mesurable? while implementing I didn't notice a slowdown in the rendering.
- The QR code is still being rendered in status-go.
- We would need many extra effort to create a second implementation of all avatars in status go that also looks as designs (wallet avatar, community, user-avatar, channel-avatar, account-avatar), why to do it? we already have a workling implementation in RN, and any change in designs is easy to perform, Also I think having two implementations of the same componet is not the best unless the obtained performance is significantly better


> 
> By deprecating the method which you intend to do in this PR we are essentially throwing away all of the work I have done into making QR codes via status-go.

It sounds too bad :/ It was not the intended result. The expected result was to have the QR code variants working, maybe if you are about to finish we could discard this PR.

> 
> The status-go implementation is customisable and we should attempt to move the initials generation and adding emojis on top of QR code from status-go side.

Is easy to add initials generation and emojis that match designs? as I said before and in this PR's description, we already have components doing it, if rendering time is not significantly better, why not just reuse them?

**Comment by J-Son89** [comment]
> > Hi @ulisesmac : Thanks for your hard work on this PR and I would like to block this PR.
> > I know the existing designs in develop do not match the designs but I have a PR open which addresses those issues : #16257 I know its a bit old ...
> 
> Oh, that PR seems stuck 🤔
> 
> > We should not move this processing over to client side simply because its much more efficient to have this on status-go side.
> 
> 🤔 IMO, unless the rendering takes a significant amount of time, it's ok to have it in CLJS side. the reasons why I think it's ok are:
> 
> * All the app components (even the more complex ones) are being rendered using RN
> * The part I'm rendering in RN is pretty simple, just a view wrapper and an overlay with an avatar
> * Unless we need that performance improvement you mention, I think we shouldn't try to optimize it. BTW, is this performnace mesurable? while implementing I didn't notice a slowdown in the rendering.
> * The QR code is still being rendered in status-go.
> * We would need many extra effort to create a second implementation of all avatars in status go that also looks as designs (wallet avatar, community, user-avatar, channel-avatar, account-avatar), why to do it? we already have a workling implementation in RN, and any change in designs is easy to perform, Also I think having two implementations of the same componet is not the best unless the obtained performance is significantly better
> 
> > By deprecating the method which you intend to do in this PR we are essentially throwing away all of the work I have done into making QR codes via status-go.
> 
> It sounds too bad :/ It was not the intended result. The expected result was to have the QR code variants working, maybe if you are about to finish we could discard this PR.
> 
> > The status-go implementation is customisable and we should attempt to move the initials generation and adding emojis on top of QR code from status-go side.
> 
> Is easy to add initials generation and emojis that match designs? as I said before and in this PR's description, we already have components doing it, if rendering time is not significantly better, why not just reuse them?

imo I tend to agree with @ulisesmac, is this something we really need to optimise that much? 🤔 Albeit that is just my general sense and I have not looked into this so much.

@cammellos I'm curious to hear your thoughts on this 🤨

**Comment by siddarthkay** [comment]
Hi @J-Son89, @cammellos  
The Image server is customisable and has the capabilities to add stuff like an emoji on top of a QR or an avatar with a background the same way it is shown in screenshots above in this PR.
I believe @yqrashawn also made few efforts on generating something similar for some other feature.

My only point here is that we have done work to superImpose user's profile picture on top of QR all generated from status-go side and with this PR we are moving away from that approach, If we want to do that then there is no point making a QR code from status-go, we might as well do everything client side?

Back when I was implementing this feature I looked into many issues before working on the approach to render a QR code via the media server and that is the reason why I am against this PR, we should at least have a conversation before abandoning this approach. 
cc @Samyoul

**Comment by siddarthkay** [comment]
note: when I say it has capabilities I mean the image uri can take in more query parameters and we will have to write code in status-go to produce the output.. it is not ready ready, what I mean is that we should have that business logic in status-go side.

**Comment by J-Son89** [comment]
Yep I agree @siddarthkay - perhaps better to move this conversation to the chat or a GitHub discussion? Or maybe a call with the relevant people is quicker to get an approach agreed upon 👍

**Comment by ulisesmac** [comment]
Thanks everyone for taking time reviewing this PR.

This PR is being closed since the team decided that status-go is a better place for this feature.
@siddarthkay Could you please add a link to the issue in status-go? thanks

**Comment by ulisesmac** [comment]
Thanks everyone for taking time reviewing this PR.

This PR is being closed since the team decided that status-go is a better place for this feature.
@siddarthkay Could you please add a link to the issue in status-go? thanks

**Comment by yqrashawn** [comment]
There's `ExtractInitials` in status-go at https://github.com/status-im/status-go/blob/2afe5a269de11bad5412a64e27d33bd49a28d588/images/initials.go#L21

File: src/utils/image_server.cljs
Comment (ulisesmac) [reply src/utils/image_server.cljs]: @siddarthkay  The deprecation has been removed

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (ulisesmac) [reply src/quo2/components/avatars/wallet_user_avatar.cljs:45]: @J-Son89 This component is not using our `with-theme` fn, also it'd be good to refactor it's API sizes and the `circle-color` var is not changing according to the theme (it's always 50).

Instead of fixing it here, I created an issue:

- #17481

File: src/quo2/components/share/qr_code/view.cljs
Comment (ulisesmac) [reply src/quo2/components/share/qr_code/view.cljs:23]: yes o_o
image

File: src/status_im2/contexts/chat/messages/content/text/view.cljs
Comment (ulisesmac) [reply src/status_im2/contexts/chat/messages/content/text/view.cljs]: removed

File: src/quo2/components/avatars/wallet_user_avatar.cljs
Comment (J-Son89) [reply src/quo2/components/avatars/wallet_user_avatar.cljs:45]: Sure sounds good to me. Thanks @ulisesmac!

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (ulisesmac) [reply src/status_im2/contexts/quo_preview/share/qr_code.cljs]: Done! good catch!

File: src/status_im2/contexts/quo_preview/share/qr_code.cljs
Comment (ulisesmac) [reply src/status_im2/contexts/quo_preview/share/qr_code.cljs:98]: oh yes, I totally agree, this is due to the channel-avatar implementation:
(ns: `quo2.components.avatars.channel-avatar.style`)
image

I think we should standarize the color mechanism according to your idea :+1: 

About the comment, I think the code is a bit weird because of this special case, so I decided to explan a little more about it.

Comment (J-Son89) [reply src/status_im2/contexts/quo_preview/share/qr_code.cljs:98]: Btw the custom color function will currently take a hex and use that 👍 but yeah we need to update that component directly.

**Comment by ulisesmac** [comment]
@siddarthkay This PR has been reopen and updated as discussed.

@J-Son89  I replied to your comments and pushed some commits addressing them. thanks for the review

**Review by J-Son89 — APPROVED** [review]

**Review by siddarthkay — CHANGES_REQUESTED** [review]
Hi @ulisesmac : can you please move the QR code preview code which uses the media server  to another file instead of deleting it
It should still be useable.
 
I don't have a strong preference on what it could be called or where it lives just that it should still be accessible.

**Comment by ulisesmac** [comment]
> Hi @ulisesmac : can you please move the QR code preview code which uses the media server to another file instead of deleting it It should still be useable.
> 
> I don't have a strong preference on what it could be called or where it lives just that it should still be accessible.

Hi @siddarthkay 
Do you mean, keep two preview screens for the QR code?
I don't get what exactly you are asking, the current implementation uses the media-server to show the QR code, the avatar is drawn on top of that image.

it'd be good If you could give me a quick explanation of the namespaces and content that you are willing to have 👍

**Comment by siddarthkay** [comment]
Hi @ulisesmac 
Sorry for the confusion, let me try to explain in bit more detail : 

What I mean is that we have this screen in quo component preview in `develop`

Simulator Screenshot - iPhone 11 Pro - 2023-10-03 at 19 17 45

This screen allowed us to play around with media server url and see the resulting QR implementation.
This has now been replaced with the client side implementation in your PR
My ask is to leave this preview screen as it is and have another one for client side QR code variants.
OR
you can keep yours as the default and move mine under media-server QR code.
Whichever namespace you wish to choose is up to you, I don't have a strong preference.
My ask is to not delete the existing preview since it is useful and since we may move the QR variants from client side to status-go in the future.

**Comment by VolodLytvynenko** [comment]
Hello @ulisesmac, I have a question. Is it possible to set the avatar as a square for now? I see [here](https://github.com/status-im/status-mobile/pull/17221#issue-1886710223) you did it or is this just a screenshot from some Figma screen? :)
image

**Comment by VolodLytvynenko** [comment]
@ulisesmac Take a look, please. One issue is found

## ISSUE 1: The non-revealed  QR code does not match the design

### Steps:
1. Go to Syncing -> tap [+] 

### Actual result:
The non-revealed QR code does not match the design. It is shown as follows:
image

### Expected result:
The non-revealed QR code should match the design, which is hidden behind a lighter curtain, as shown below:

image
https://www.figma.com/file/V6nlpAWIf2e1XU8RJ9yQPe/Syncing-for-Mobile?type=design&node-id=675-177939&mode=design&t=cqYVBwMlprhlJjSJ-0

**Comment by ulisesmac** [comment]
> Hi @ulisesmac Sorry for the confusion, let me try to explain in bit more detail :
> 
> What I mean is that we have this screen in quo component preview in `develop`
> 
>
> This screen allowed us to play around with media server url and see the resulting QR implementation. This has now been replaced with the client side implementation in your PR My ask is to leave this preview screen as it is and have another one for client side QR code variants. OR you can keep yours as the default and move mine under media-server QR code. Whichever namespace you wish to choose is up to you, I don't have a strong preference. My ask is to not delete the existing preview since it is useful and since we may move the QR variants from client side to status-go in the future.

@siddarthkay 

You can keep testing the status-go media server implementation in the current preview screen :)
Just pick the "None" variant, and the result is the same as before. The only missing thing is the error correction level, wich previously was a parameter, but John said we will use the highest correction level, so we can skip that parameter :)

**Comment by ulisesmac** [comment]
> Hello @ulisesmac, I have a question. Is it possible to set the avatar as a square for now? I see [here](https://github.com/status-im/status-mobile/pull/17221#issue-1886710223) you did it or is this just a screenshot from some Figma screen? :) image

Hey @VolodLytvynenko 
No, it's not possible to get the QR codes as a square in the app. I took a screenshot from the app  preview screen and  cropped it, just to show them in this PR description.

**Comment by ulisesmac** [comment]
> @ulisesmac Take a look, please. One issue is found
> ## ISSUE 1: The non-revealed QR code does not match the design
> ### Steps:
> 
>     1. Go to Syncing -> tap [+]
> 
> 
> ### Actual result:
> 
> The non-revealed QR code does not match the design. It is shown as follows: image
> ### Expected result:
> 
> The non-revealed QR code should match the design, which is hidden behind a lighter curtain, as shown below:
> 
> image https://www.figma.com/file/V6nlpAWIf2e1XU8RJ9yQPe/Syncing-for-Mobile?type=design&node-id=675-177939&mode=design&t=cqYVBwMlprhlJjSJ-0

Hey @VolodLytvynenko 

Sorry, this screen had been updated in a previous PR and I broke it when rebased. Now it's fixed :+1:

**Comment by VolodLytvynenko** [comment]
Hi @ulisesmac thank you for the fix. Sorry, I have missed one low issue which might be fixed in a separate follow-up. The decision is up to you

## ISSUE 2:  Incorrect avatar color in profile QR

### Steps:
1. Don't set up image for avatar
2. Open profile QR image
3. Check the avatar

### Actual result:
The avatar has a dark background with light-colored letters inside
image

### Expected result:
The avatar should have a light background with dark-colored letters inside
image

**Comment by VolodLytvynenko** [comment]
Hi @Francesca-G, could you check PR and [this](https://github.com/status-im/status-mobile/pull/17221#issuecomment-1748732434) issue? In most designs, the avatar has a dark background with light-colored letters, but in the current design at [link](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=28187-85679&mode=design&t=l5AnjJypiQASDoFT-0), it has a light background with dark-colored letters. Is this intentional or a mistake?

**Review by Francesca-G — APPROVED** [review]
Looks good ✨

@VolodLytvynenko the component we're using in the QR codes is the Wallet User Avatar component ([here](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=114%3A7646&mode=design&t=MXDpYBvwFmw5hCL6-1)) which has light background and dark letters.

I think that the component you see in most designs is the User Avatar component ([here](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=115%3A6787&mode=design&t=MXDpYBvwFmw5hCL6-1)) which has dark background and light letters.

If you have any more doubts I can check with the rest of the team :)

File: package.json
Comment (siddarthkay) [inline package.json:38]: Hey @ulisesmac : I think you may have to also check in the `yarn.lock` for this corresponding change.

**Review by siddarthkay — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
> I think that the component you see in most designs is the User Avatar component ([here](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=115%3A6787&mode=design&t=MXDpYBvwFmw5hCL6-1)) which has dark background and light letters.
> 
> If you have any more doubts I can check with the rest of the team :)

@Francesca-G, I see it now. My apologies for missing that wallet [avatar design](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=114-7646&mode=design). Thank you for providing link! From what I gather, the issue mentioned at https://github.com/status-im/status-mobile/pull/17221#issuecomment-1748732434 doesn't seem to be an actual issue. It appears that the QR code I checked is related to the user profile QR, not the wallet. 
Please correct me if I'm mistaken. In my view, the PR is ready for merging.

**Pull request merged**
